File: src/spirite/base/brains/HotkeyManager.java

package spirite.base.brains;

import java.awt.event.InputEvent;
import java.awt.event.KeyEvent;
import java.io.UnsupportedEncodingException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.prefs.Preferences;

import org.apache.commons.collections4.BidiMap;
import org.apache.commons.collections4.bidimap.DualHashBidiMap;


/***
 * Each hotkey-able action has a text identifier of the form
 * {context}.{action}
 * 
 * Not just a visual significance, the context text will be used to determine which
 * component needs to handle the hotkey command.  The hotkey manager stores all
 * links between the hotkey-able actions and their corresponding Hotkey (a hotkey
 * is a key-code and its modifier, including possibly some extended modifiers).
 * 
 * The HotkeyManager does not actually dictate the actions to the corresponding
 * component (that's handled in the Main Frame), it simply stores all the hotkeys
 * and what they are linked to.
 * 
 * TODO: This entire class really needs to be re-written such that the following
 * functions are allowed and work as intended:
 * -Allowing multiple hotkeys for a single command (but not vice versa)
 * -Allow for saving and loading of Hotkey maps (both to preference string and
 * to file) without using that ugly ISO encoding.
 * 
 * @author Rory Burks
 *
 */
public class HotkeyManager {
    private final Preferences prefs;    
    private final BidiMap<Hotkey,String> hotkey_map;
    

    HotkeyManager() {
        prefs = Preferences.userNodeForPackage(spirite.pc.Spirite.class);
        hotkey_map = new DualHashBidiMap<>();

        loadHotkeys(
            new Object[][] {
                {"context.zoom_in", (new Hotkey( KeyEvent.VK_ADD, 0))},
                {"context.zoom_out", (new Hotkey( KeyEvent.VK_SUBTRACT, 0))},
                {"context.zoom_in_slow", (new Hotkey( KeyEvent.VK_ADD, InputEvent.CTRL_DOWN_MASK))},
                {"context.zoom_out_slow", (new Hotkey( KeyEvent.VK_SUBTRACT, InputEvent.CTRL_DOWN_MASK))},
                {"context.zoom_0", (new Hotkey( KeyEvent.VK_NUMPAD0, InputEvent.CTRL_DOWN_MASK))},

                {"toolset.PEN", (new Hotkey( KeyEvent.VK_P, 0))},
                {"toolset.ERASER", (new Hotkey( KeyEvent.VK_E, InputEvent.SHIFT_DOWN_MASK))},
                {"toolset.FILL", (new Hotkey( KeyEvent.VK_B, InputEvent.SHIFT_DOWN_MASK))},
                {"toolset.BOX_SELECTION", (new Hotkey( KeyEvent.VK_R, 0))},
                {"toolset.MOVE", (new Hotkey( KeyEvent.VK_M, 0))},
                {"toolset.COLOR_PICKER", (new Hotkey( KeyEvent.VK_O, 0))},
                {"toolset.PIXEL", (new Hotkey( KeyEvent.VK_A, 0))},
                {"toolset.COMPOSER", (new Hotkey( KeyEvent.VK_CAPS_LOCK, 0))},

                {"palette.swap", (new Hotkey( KeyEvent.VK_X, 0))},
                {"palette.swapBack", (new Hotkey( KeyEvent.VK_Z, 0))},
                
                {"draw.newLayerQuick", (new Hotkey( KeyEvent.VK_INSERT, 0))},
                {"draw.undo", (new Hotkey( KeyEvent.VK_Z, InputEvent.CTRL_DOWN_MASK))},
                {"draw.redo", (new Hotkey( KeyEvent.VK_Z, InputEvent.CTRL_DOWN_MASK | InputEvent.SHIFT_DOWN_MASK))},
                {"draw.clearLayer", (new Hotkey( KeyEvent.VK_DELETE, 0))},
                {"draw.invert", (new Hotkey( KeyEvent.VK_I, InputEvent.CTRL_DOWN_MASK ))},
                {"draw.toggle_reference", (new Hotkey( KeyEvent.VK_BACK_QUOTE, 0))},	
                {"draw.lift_to_reference", (new Hotkey( KeyEvent.VK_F, InputEvent.CTRL_DOWN_MASK | InputEvent.SHIFT_DOWN_MASK))},	

                {"draw.shiftLeft", (new Hotkey( KeyEvent.VK_LEFT, InputEvent.SHIFT_DOWN_MASK | InputEvent.CTRL_DOWN_MASK))},
                {"draw.shiftRight", (new Hotkey( KeyEvent.VK_RIGHT, InputEvent.SHIFT_DOWN_MASK | InputEvent.CTRL_DOWN_MASK))},
                {"draw.shiftDown", (new Hotkey( KeyEvent.VK_DOWN, InputEvent.SHIFT_DOWN_MASK | InputEvent.CTRL_DOWN_MASK))},
                {"draw.shiftUp", (new Hotkey( KeyEvent.VK_UP, InputEvent.SHIFT_DOWN_MASK | InputEvent.CTRL_DOWN_MASK))},

                {"select.all", (new Hotkey( KeyEvent.VK_A, InputEvent.CTRL_DOWN_MASK))},
                {"select.none", (new Hotkey( KeyEvent.VK_D, InputEvent.CTRL_DOWN_MASK | InputEvent.SHIFT_DOWN_MASK))},
                {"select.invert", (new Hotkey( KeyEvent.VK_I, InputEvent.CTRL_DOWN_MASK | InputEvent.SHIFT_DOWN_MASK))},
                
                {"global.save_image", (new Hotkey( KeyEvent.VK_S, InputEvent.CTRL_DOWN_MASK))},
                {"global.copy", (new Hotkey( KeyEvent.VK_C, InputEvent.CTRL_DOWN_MASK))},
                {"global.copyVisible", (new Hotkey( KeyEvent.VK_C, InputEvent.CTRL_DOWN_MASK | InputEvent.SHIFT_DOWN_MASK))},
                {"global.paste", (new Hotkey( KeyEvent.VK_V, InputEvent.CTRL_DOWN_MASK))},
                {"global.pasteAsLayer", (new Hotkey( KeyEvent.VK_V, InputEvent.CTRL_DOWN_MASK | InputEvent.SHIFT_DOWN_MASK))},
                {"global.cut", (new Hotkey( KeyEvent.VK_X, InputEvent.CTRL_DOWN_MASK))},
                
                // TODO: This should really be anim., but that might require restructuring/rethinking of
                //	command execution system.
                {"draw.addGapQuick", new Hotkey( KeyEvent.VK_INSERT, InputEvent.SHIFT_DOWN_MASK)},

                {"global.debug1", (new Hotkey( KeyEvent.VK_1, InputEvent.CTRL_DOWN_MASK | InputEvent.SHIFT_DOWN_MASK))},
            }
        );
        
        loadBinds(
    		new Object[][] {
    			{"stylus", "toolset.PEN"},
    			{"eraser", "toolset.ERASER"},
    			{"mouse" , "toolset.PEN"}
    		}
        );
    }
    
    /***
     * Returns whether the given keycode is considered a "Modifier" (in other
     * words that key alone should not serve as a hotkey).
     */
    public static boolean isModifier( int keycode) {
    	switch( keycode) {
    	case KeyEvent.VK_CONTROL:
    	case KeyEvent.VK_SHIFT:
    	case KeyEvent.VK_ALT:
    		return true;
		default:
			return false;
    	}
    }

    /**
     * Loads the hotkey data from preference space using a hard-coded set of key
     * commands and default hotkeys with their modifier 
     * @param data
     */
    private void loadHotkeys( Object[][] data) {
        for( Object[] obj : data) {
            Hotkey key = (Hotkey)obj[1];
            String s = prefs.get( (String)obj[0], key.toPrefString());

            key.fromPrefString(s);

            hotkey_map.put(key, (String)obj[0]);
        }
    }
    
    private void loadBinds( Object[][] data) {
    	
    }
    
    public String getCommand( int key, int modifier) {
        String s = hotkey_map.get(new Hotkey(key, modifier));
        return s;
    }

    public Hotkey getHotkey( String command) {
        return hotkey_map.getKey(command);
    }
    
    public void setCommand( int key, int modifier, String command) {
    	hotkey_map.put( new Hotkey( key, modifier), command);
    }
    

    /**
     * A Hotkey
     */
    public static class Hotkey
    {
        int key;        // from MouseEvent.getButton
                        // or   KeyEvent.getKey
        int modifier;   // from InputEvent.getModifiersEx

        public Hotkey( int key, int modifier) {
            this.key = key;
            this.modifier = modifier;
        }

        public String toPrefString() {
            ByteBuffer buff = ByteBuffer.allocate(10);
            buff.order(ByteOrder.BIG_ENDIAN);

            buff.putInt(key);
            buff.putInt(modifier);

            // !!! Bad
            try {
            	String str = new String(buff.array(), "ISO-8859-1");
                return new String(buff.array(), "ISO-8859-1");//buff.array().toString();
            } catch (UnsupportedEncodingException ex) {
                return "";
            }
        }

        public void fromPrefString( String s) {
            ByteBuffer buff;

            // !!! Bad
            try {
                buff = ByteBuffer.wrap(s.getBytes("ISO-8859-1"));
            } catch (UnsupportedEncodingException ex) {
                buff = ByteBuffer.allocate(0);
            }
            buff.order( ByteOrder.BIG_ENDIAN);

            key = buff.getInt();
            modifier = buff.getInt();
        }

        @Override
        public String toString() {

            return ((modifier != 0) ? KeyEvent.getModifiersExText(modifier) + " + " : "") + KeyEvent.getKeyText(key);
        }

        // Added to make sure the HashMap works
        @Override
        public boolean equals( Object obj) {
            if( obj instanceof Hotkey) {
                Hotkey other = (Hotkey) obj;

                if( key == other.key && modifier == other.modifier)
                    return true;
            }
            return false;
        }

        // !!! Automaticall generated by Netbeans
        @Override
        public int hashCode() {
            int hash = 7;
            hash = 23 * hash + this.key;
            hash = 23 * hash + this.modifier;
            return hash;
        }
    }
}

File: src/main/kotlin/spirite/base/brains/HotkeyManager.kt
package spirite.base.brains

import rb.extendo.dataStructures.MutableOneToManyMap
import spirite.base.brains.commands.*
import spirite.base.brains.commands.DrawCommandExecutor.DrawCommand
import spirite.base.brains.commands.IsolationCommandExecutor.IsolationCommand
import spirite.base.brains.commands.ToolsetCommandExecutor.ToolCommand
import spirite.base.brains.commands.WorkViewCommandExecutor.WorkViewCommand
import spirite.base.brains.settings.IPreferences
import java.awt.event.InputEvent
import java.awt.event.KeyEvent

data class Hotkey(
        val key: Int,
        val modifier: Int
)

class KeyCommand(
        val commandString : String,
        val objectCreator: ((IMasterControl)->Any?)?  = null
)
{
    // region generated
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as KeyCommand
        if (commandString != other.commandString) return false
        return true
    }

    override fun hashCode(): Int {
        return commandString.hashCode()
    }
    // endregion
}

interface IHotkeyManager {
    fun getCommand( key: Hotkey) : KeyCommand?
    fun getHotkeysForCommand( command: String) : List<Hotkey>?

    fun setCommand( key: Hotkey, command : KeyCommand)
    fun removeHotkey( key: Hotkey)
}

private const val ALT = InputEvent.ALT_DOWN_MASK
private const val CTRL = InputEvent.CTRL_DOWN_MASK
private const val SHIFT = InputEvent.SHIFT_DOWN_MASK

private val defaultHotkeys = mapOf(
        WorkViewCommand.ZOOM_IN.keyCommand to (Hotkey( KeyEvent.VK_ADD, 0)),
        WorkViewCommand.ZOOM_OUT.keyCommand to (Hotkey( KeyEvent.VK_ADD, CTRL or SHIFT)),
        WorkViewCommand.ZOOM_IN_SLOW.keyCommand to (Hotkey( KeyEvent.VK_ADD, CTRL)),
        WorkViewCommand.ZOOM_OUT_SLOW.keyCommand to (Hotkey( KeyEvent.VK_SUBTRACT, CTRL)),
        WorkViewCommand.ZOOM_0.keyCommand to (Hotkey( KeyEvent.VK_NUMPAD0, CTRL)),

        ToolCommand.Pen.keyCommand to (Hotkey( KeyEvent.VK_B, 0)),
        ToolCommand.Eraser.keyCommand to (Hotkey( KeyEvent.VK_E, 0)),
        ToolCommand.Fill.keyCommand to (Hotkey( KeyEvent.VK_B, SHIFT)),
        ToolCommand.ShapeSelection.keyCommand to (Hotkey( KeyEvent.VK_R, 0)),
        ToolCommand.Move.keyCommand to (Hotkey( KeyEvent.VK_T, 0)),
        ToolCommand.ColorPicker.keyCommand to (Hotkey( KeyEvent.VK_O, 0)),
        ToolCommand.Pixel.keyCommand to (Hotkey( KeyEvent.VK_A, 0)),
        ToolCommand.Rigger.keyCommand to (Hotkey( KeyEvent.VK_R, SHIFT)),
        ToolCommand.MagneticFill.keyCommand to (Hotkey(KeyEvent.VK_F, 0)),
        ToolCommand.FreeSelection.keyCommand to (Hotkey(KeyEvent.VK_F, SHIFT)),

        ToolCommand.SetMode_1.keyCommand to (Hotkey(KeyEvent.VK_1, SHIFT)),
        ToolCommand.SetMode_2.keyCommand to (Hotkey(KeyEvent.VK_2, SHIFT)),
        ToolCommand.SetMode_3.keyCommand to (Hotkey(KeyEvent.VK_3, SHIFT)),
        ToolCommand.SetMode_4.keyCommand to (Hotkey(KeyEvent.VK_4, SHIFT)),

        ToolCommand.DecreasePenSize.keyCommand to (Hotkey( KeyEvent.VK_OPEN_BRACKET, 0)),
        ToolCommand.IncreasePenSize.keyCommand to (Hotkey( KeyEvent.VK_OPEN_BRACKET, CTRL or SHIFT)),

        PaletteCommands.Swap.keyCommand to (Hotkey( KeyEvent.VK_X, 0)),
        PaletteCommands.SwapBack.keyCommand to (Hotkey( KeyEvent.VK_Z, 0)),
        PaletteCommands.SwitchModes.keyCommand to (Hotkey(KeyEvent.VK_P, SHIFT or CTRL)),
        PaletteCommands.CyclePalettes.keyCommand to (Hotkey(KeyEvent.VK_P, CTRL)),

        NodeCommands.QuickNewLayer.keyCommand to (Hotkey( KeyEvent.VK_INSERT, 0)),
        NodeCommands.NewSpriteLayer.keyCommand to Hotkey( KeyEvent.VK_INSERT, SHIFT),
        NodeCommands.NewGroup.keyCommand to (Hotkey(KeyEvent.VK_INSERT, CTRL)),
        NodeCommands.NewPuppetLayer.keyCommand to (Hotkey(KeyEvent.VK_INSERT, CTRL or SHIFT)),
        NodeCommands.MoveDown.keyCommand to (Hotkey(KeyEvent.VK_DOWN, CTRL)),
        NodeCommands.MoveUp.keyCommand to (Hotkey(KeyEvent.VK_UP, CTRL)),
        NodeCommands.Duplicate.keyCommand to (Hotkey(KeyEvent.VK_U, CTRL)),

        WorkspaceCommands.ToggleView.keyCommand to (Hotkey(KeyEvent.VK_TAB,0)),
        WorkspaceCommands.ResetOtherView.keyCommand to (Hotkey(KeyEvent.VK_TAB, SHIFT)),
        WorkspaceCommands.ResetOtherView.keyCommand to (Hotkey(KeyEvent.VK_TAB, CTRL)),

        IsolationCommand.TOGGLE_ISOLATION.keyCommand to (Hotkey(KeyEvent.VK_BACK_QUOTE, 0)),
        IsolationCommand.ISOLATE_LAYER.keyCommand to (Hotkey(KeyEvent.VK_BACK_QUOTE, CTRL)),
        IsolationCommand.CLEAR_ALL_ISOLATION.keyCommand to (Hotkey(KeyEvent.VK_BACK_QUOTE, CTRL or SHIFT)),

        DrawCommand.UNDO.keyCommand to (Hotkey( KeyEvent.VK_Z, CTRL)),
        DrawCommand.REDO.keyCommand to (Hotkey( KeyEvent.VK_Z, CTRL or SHIFT)),
        DrawCommand.CLEAR.keyCommand to (Hotkey( KeyEvent.VK_DELETE, 0)),
        DrawCommand.INVERT.keyCommand to (Hotkey( KeyEvent.VK_I, CTRL )),
        //KeyCommand("draw.toggle_reference") to (Hotkey( KeyEvent.VK_BACK_QUOTE, 0)),
        KeyCommand("draw.lift_to_reference") to (Hotkey( KeyEvent.VK_F, CTRL or SHIFT)),

        DrawCommand.SHIFT_LEFT.keyCommand to (Hotkey( KeyEvent.VK_LEFT, SHIFT or CTRL)),
        DrawCommand.SHIFT_RIGHT.keyCommand to (Hotkey( KeyEvent.VK_RIGHT, SHIFT or CTRL)),
        DrawCommand.SHIFT_DOWN.keyCommand to (Hotkey( KeyEvent.VK_DOWN, SHIFT or CTRL)),
        DrawCommand.SHIFT_UP.keyCommand to (Hotkey( KeyEvent.VK_UP, SHIFT or CTRL)),
        DrawCommand.SCALE3x.keyCommand to (Hotkey( KeyEvent.VK_F3, SHIFT)),


        SelectCommand.All.keyCommand to (Hotkey( KeyEvent.VK_A, CTRL)),
        SelectCommand.None.keyCommand to (Hotkey( KeyEvent.VK_D, CTRL or SHIFT)),
        SelectCommand.Invert.keyCommand to (Hotkey( KeyEvent.VK_I, CTRL or SHIFT)),
        SelectCommand.LiftInPlace.keyCommand to (Hotkey( KeyEvent.VK_X, SHIFT)),

        GlobalCommands.SaveWorkspace.keyCommand to (Hotkey( KeyEvent.VK_S, CTRL)),
        GlobalCommands.Copy.keyCommand to (Hotkey( KeyEvent.VK_C, CTRL)),
        GlobalCommands.CopyVisible.keyCommand to (Hotkey( KeyEvent.VK_C, CTRL or SHIFT)),
        GlobalCommands.Paste.keyCommand to (Hotkey( KeyEvent.VK_V, CTRL)),
        GlobalCommands.PasteAsLayer.keyCommand to (Hotkey( KeyEvent.VK_V, CTRL or SHIFT)),
        GlobalCommands.Cut.keyCommand to (Hotkey( KeyEvent.VK_X, CTRL)),
        GlobalCommands.Open.keyCommand to (Hotkey( KeyEvent.VK_O, CTRL)),
        GlobalCommands.NewWorkspace.keyCommand to (Hotkey( KeyEvent.VK_N, CTRL)),

        // DEBUG
        GlobalCommands.PurgeUndoHistory.keyCommand to (Hotkey(KeyEvent.VK_F11, ALT or SHIFT)),
        GlobalCommands.CopyAllLayer.keyCommand to (Hotkey(KeyEvent.VK_F12, CTRL or SHIFT)),

        // TODO: This should really be animation., but that might require restructuring/rethinking of
        //	command execution system.

        KeyCommand("global.debug1") to (Hotkey( KeyEvent.VK_1, CTRL or SHIFT)),

        KeyCommand("frame.focus.LAYER_PROPERTIES") to (Hotkey(KeyEvent.VK_R, SHIFT)))

class HotkeyManager(
        val preferences: IPreferences
) : IHotkeyManager {
    private val hotkeyMap = MutableOneToManyMap<KeyCommand, Hotkey>()

    init {
        // Load either Default or saved preferences
        defaultHotkeys.entries.forEach {
            val command = it.key
            val defaultHotkey = it.value

            val preferenceString = preferences.getString(command.commandString)
            if( preferenceString == null) {
                hotkeyMap.assosciate(defaultHotkey, command)
            }
            else if( !preferenceString.isBlank()) {
                preferenceStringToHotkeys(preferenceString)
                        .forEach { hotkey -> hotkeyMap.assosciate(hotkey, command) }
            }
        }
    }

    override fun getCommand(key: Hotkey): KeyCommand? = hotkeyMap.getOne(key)
    override fun getHotkeysForCommand(command: String): List<Hotkey>? = hotkeyMap.getMany(KeyCommand(command))
    override fun setCommand(key: Hotkey, command: KeyCommand) {
        val oldCommand = hotkeyMap.getOne(key)

        hotkeyMap.assosciate(key, command)

        // Change Old
        if( oldCommand != null) {
            val oldMany = hotkeyMap.getMany(oldCommand)
            if( oldMany == null)
                preferences.putString(oldCommand.commandString,"")
            else
                preferences.putString(oldCommand.commandString, hotkeysToPreferenceString(oldMany))
        }

        // Change New
        val newCommandsRewritten = hotkeyMap.getMany(command)
        if( newCommandsRewritten != null)
            preferences.putString(command.commandString, hotkeysToPreferenceString(newCommandsRewritten))

    }
    override fun removeHotkey(key: Hotkey) {
        val command = hotkeyMap.many_to_one[key]
        hotkeyMap.dissociate(key)

        if( command != null) {
            val hotkeysToRewrite = hotkeyMap.getMany( command)
            if( hotkeysToRewrite == null)
                preferences.putString(command.commandString,"")
            else
                preferences.putString(command.commandString, hotkeysToPreferenceString(hotkeysToRewrite))
        }
    }

    private fun hotkeysToPreferenceString( keys: List<Hotkey>) : String {
        return keys.map { "${it.key},${it.modifier}" }
                .joinToString (";")
    }

    private fun preferenceStringToHotkeys( prefString : String) : List<Hotkey> {
        return prefString.split(";")
                .map {
                    val params = it.split(",")
                    Hotkey( params[0].toInt(10), params[1].toInt(10))
                }
    }
}

